### PR TITLE
Show remote carets

### PR DESCRIPTION
**Pull request by Worie**
Here we go. I tried to follow your codestyle, I guess it should be ok. The changes provided should cover every input from peers (multiple carets, removals, additions, pastes and cuts)

If this PR's alright, then in the future I guess we should create a html template for carets (for future features such as displaying the carets owner name and for code maintenance purposes)

This solves #22

### CODE REVIEW

**Comment by Worie** [comment]
Just realized that it works well when there are only two peers, but fails with more than that.

The problem is that upon recieving the change event, the `remoteCarets` set is being cleared.

So in scenario: 

1. Peer A connects
2. Peer B connects
3. Peer C connects
4. Peer B moves his caret
5. Peer A and C sees it
6. Peer C moves his caret.
7. Peer A loses the state of peer B caret.

So, each peer should have probably his own `Set()`.

File: main.js [hunk 1]
```diff
@@ -232,15 +233,17 @@ define(function (require, exports, module) {
...
+      oldEditor._codeMirror.off('beforeSelectionChange', onBeforeSelectionChange)
     }
 
     if (newEditor) {
       currentEditor = newEditor
       documentRelativePath = FileUtils.getRelativeFilename(projectBasePath, newEditor.document.file.fullPath)
-      newEditor._codeMirror.on('change', sendLocalChange)
-      newEditor._codeMirror.on('beforeSelectionChange', sendLocalSelect)
+      newEditor._codeMirror.on('change', onCodeChange)
+      newEditor._codeMirror.on('changes', onCodeChanges)
```
Comment (t-mullen) [inline main.js]: If we are going to start listening to operation-batched changes, we should only use the 'changes' event.

File: main.js [hunk 2]
```diff
@@ -254,9 +257,60 @@ define(function (require, exports, module) {
...
+  function onBeforeSelectionChange (cm, change) {
+    sendLocalSelect(cm, change)
+    sendLocalCaretsMoves(cm, change)
+  }
+  
+  function isSelection (range) {
+    return range.head.ch !== range.anchor.ch || range.head.line !== range.anchor.line
+  }
+  
+  function sendLocalCaretsMoves (cm, change) {
```
Comment (t-mullen) [inline main.js:274]: Do you mind adding a few comments to this to explain how you are calculating the caret positions from the changes?

Comment (Worie) [reply main.js:274]: Sure, I'll sit to it when finish your review :)

File: main.js [hunk 3]
```diff
@@ -396,13 +450,30 @@ define(function (require, exports, module) {
     })
   }
   
+  function handleRemoteCarets (data) {
+    if (fromWebPath(data.filePath) !== documentRelativePath) return
```
Comment (t-mullen) [inline main.js]: Fine for now, but add a TODO so we can support tracking carets in closed files. (Needs to be done for selections to if I remember correctly.)

**Comment by t-mullen** [comment]
Looks good, thanks for figuring this out. I added a few comments.

File: main.js [hunk 4]
```diff
@@ -34,6 +34,7 @@ define(function (require, exports, module) {
   var nickname = null
   var knownDocs = {}
   var room
+  var remoteCarets = new Set()
```
Comment (t-mullen) [inline main.js]: I don't use ES6 anywhere else. Please use an object set and use keys to iterate.

**Review by t-mullen — CHANGES_REQUESTED** [review]
Changed my mind. YJS integration will take too long.

Please make the requested style changes and I'll run some tests.

**Comment by Worie** [comment]
Guess that's it. Besides the requested changes I've left a comment about the issue I mentioned in https://github.com/RationalCoding/multihack-brackets/pull/23#issuecomment-298181615 earlier.

**Comment by t-mullen** [comment]
OK, seems to be working great aside from the issue with >2 users. I'll fix that in a later commit when I do cursor colours. Thanks @Worie !

Would you like to be added as a maintainer? I can always use some more help developing this :)

**Comment by Worie** [comment]
I'm glad it's ok :) When it comes to me as a maintainer - I can't really promise anything as I don't know how much spare time will I have, but sure - that'd be nice, thanks!

**Pull request merged**